data_mgr.py

```diff
@@ -0,0 +1,238 @@
+        return (w_id, None)
+    else:
+        ws_data = METHOD_MAP[method]()
+        ws_data.ParseFromString(pb_msg)
+    return (w_id, ws_data)
+
+
+class DataManager(object):
+
+    def __init__(self, ws_mgr):
+        self.ws_mgr = ws_mgr
+        self.data = {}
+
+    # **
+    # Data syncing
+    # **
+    async def entire_workflow_update(self, ids=None):
+        if ids is None:
+            ids = []
+        new_data = {}
+        ws_args = (
+            (w_id, info['req_client'], 'pb_entire_workflow')
+            for w_id, info in self.ws_mgr.workflows.items())
+        items = await gen.multi(
+            get_workflow_data(*args) for args in ws_args
+            if not ids or args[0] in ids)
+        for w_id, result in items:
+            if result is not None and result != MSG_TIMEOUT:
+                new_data[w_id] = result
+        if not ids:
+            # atomic update
+            self.data = new_data
+        else:
+            # apparently faster than update(new_data)
+            for w_id, flow in new_data.items():
+                self.data[w_id] = flow
+
+    # **
+    # Data access filters
+    # **
+    def get_workflow_msgs(self, args):
+        """Return list of workflows."""
+        result = []
+        for flow_msg in self.data.values():
+            if self._workflow_filter(flow_msg, args):
+                result.append(flow_msg)
+        return result
+
+    def get_nodes_all(self, node_type, args):
+        """Return list of nodes (defns/proxies/jobs)."""
+        w_args = {
+            'ids': args['workflows'],
+            'exids': args['exworkflows'],
+        }
+        nodes = [
+            n for k in self.get_workflow_msgs(w_args)
+            for n in getattr(k, node_type)]
+        result = []
+        for node in nodes:
+            if self._node_filter(node, args):
+                result.append(node)
+        return result
+
+    def get_nodes_id(self, node_type, args):
+        """Return protobuf node objects for given id."""
+        nat_ids = args.get('native_ids', [])
+        w_ids = []
+        for nat_id in nat_ids:
+            o_name, w_name, nid = nat_id.split('/', 2)
+            w_ids.append(f"{o_name}/{w_name}")
+        if node_type == 'proxy_nodes':
+            nodes = (
+                n for w_id in set(w_ids)
+                for n in (
+                    list(getattr(self.data[w_id], 'task_proxies', []))
+                    + list(getattr(self.data[w_id], 'family_proxies', []))))
+        else:
+            nodes = (
+                n for w_id in set(w_ids)
+                for n in getattr(self.data[w_id], node_type, []))
+        result = []
+        for node in nodes:
+            if ((not nat_ids or node.id in set(nat_ids)) and
+                    self._node_filter(node, args)):
+                result.append(node)
+        return result
+
+    def get_node_id(self, node_type, args):
+        """Return protobuf node object for given id."""
```

Comment:
I thought this function would return the `node_id` of a node. But instead it appears to return a node for a given node id. So I think we could have a better name here - though I am horrible with names myself. Maybe `get_node_by_id`?